include/ddc/kernels/splines/spline_evaluator_batched.hpp

```diff
@@ -0,0 +1,263 @@
+
+    KOKKOS_INLINE_FUNCTION bsplines_domain_type const bsplines_domain()
+            const noexcept // TODO : clarify name
+    {
+        return ddc::discrete_space<bsplines_type>().full_domain();
+    }
+
+    KOKKOS_INLINE_FUNCTION batch_domain_type const batch_domain() const noexcept
+    {
+        return ddc::remove_dims_of(spline_domain(), bsplines_domain());
+    }
+    template <class Layout>
+    double operator()(
+            ddc::Coordinate<IDimX...> const& coord_eval,
+            ddc::ChunkSpan<double const, spline_domain_type, Layout, memory_space> const
+                    spline_coef) const
+    {
+        return eval(coord_eval, spline_coef);
+    }
+
+    template <class Layout1, class Layout2, class Layout3, class... CoordsDims>
+    void operator()(
+            ddc::ChunkSpan<double, vals_domain_type, Layout1, memory_space> const spline_eval,
+            ddc::ChunkSpan<
+                    ddc::Coordinate<CoordsDims...> const,
+                    vals_domain_type,
+                    Layout2,
+                    memory_space> const coords_eval,
+            ddc::ChunkSpan<double const, spline_domain_type, Layout3, memory_space> const
+                    spline_coef) const
+    {
+        interpolation_domain_type interpolation_domain
+                = ddc::select<interpolation_mesh_type>(spline_eval.domain());
+        ddc::for_each(
+                ddc::policies::policy(exec_space()),
+                batch_domain(),
+                KOKKOS_CLASS_LAMBDA(typename batch_domain_type::discrete_element_type const j) {
+                    const auto spline_eval_1D = spline_eval[j];
+                    const auto coords_eval_1D = coords_eval[j];
+                    const auto spline_coef_1D = spline_coef[j];
+                    for (int i = 0; i < interpolation_domain.size(); i++) {
+                        spline_eval_1D(typename interpolation_domain_type::discrete_element_type(i))
+                                = eval(coords_eval_1D(typename interpolation_domain_type::
+                                                              discrete_element_type(i)),
+                                       spline_coef_1D);
+                    }
+                });
+    }
+
+    template <class Layout, class... CoordsDims>
+    double deriv(
+            ddc::Coordinate<CoordsDims...> const& coord_eval,
+            ddc::ChunkSpan<double const, bsplines_domain_type, Layout, memory_space> const
+                    spline_coef) const
+    {
+        return eval_no_bc<eval_deriv_type>(coord_eval, spline_coef);
+    }
+
+    template <class Layout1, class Layout2, class Layout3, class... CoordsDims>
+    void deriv(
+            ddc::ChunkSpan<double, vals_domain_type, Layout1, memory_space> const spline_eval,
+            ddc::ChunkSpan<
+                    ddc::Coordinate<CoordsDims...> const,
+                    vals_domain_type,
+                    Layout2,
+                    memory_space> const coords_eval,
```

Comment:
Is templating over `CoordDims` necessary ?

Reply:
I would not say necessary, but I do not see reasons to restrict it to `tag_type ` and I can imagine situations where it would avoid a ddc::select in the call.

Reply:
Ok there is a big subject underlying,

In current Gysela, we have successive calls to 1D interpolators, which call 1D SplineBuilder and 1D SplineEvaluator. So we never compute multidimensionnal splines in interpolators (we do 2D splines in poisson). However, there is another way to do it: a nD interpolator which calls nD SplineBuilder then nD SplineEvaluator. nD SplineBuilder is a succession of 1D SplineBuilder, nD SplineEvaluator is a succession of 1D SplineEvaluator. So, it just changes the order of the operations (BuilderX, BuilderY, EvaluatorX, EvaluatorY in place of BuilderX, EvaluatorX, BuilderY, EvaluatorY)

But there is an advantage to do so: by using an nD SplineBuilder you compute the nD Splines coef, which is a completly equivalent way to store the distribution function (basis change). And in this basis, we can do interesting things like integration on the whole space (simple transform_reduce), finite elements etc...

It may permit to mutualize operations in favorable cases, to "share" a basis change between two operators (example : you can do BuilderVx, BuilderVy, ChargeDensityComputation (integration on velocities), EvaluatorVx, EvaluatorVy to mutualize VelocityAdvection and ChargeDensityComputation). We can do even better things if we have schemes which permit to call operators in parallel (Kevin tried to perform advection and poisson in parallel in Gysela fortran).

However in the case of this question of templating over CoordsDims... it just a matter of getting a ddc::select in the call of the evaluator or not... Not a big deal, I can change it and template with tag_type if you want.

Reply:
I realize I may be a out of subject here, i'm ok with you whatever the dimensionnality of the SplineEvaluator is we can impose `ddc::Coordinate<dimensions of interest...>`, so here `ddc::Coordinate<tag_type>`

Reply:
In tests/batched_splines_evaluator.cpp, I have the following code:

`// Instantiate chunk of coordinates of dom_interpolation
    ddc::Chunk coords_eval_alloc(dom_vals, ddc::KokkosAllocator<Coord<X...>, MemorySpace>());
    ddc::ChunkSpan coords_eval = coords_eval_alloc.span_view();
    ddc::for_each(
            ddc::policies::policy(exec_space),
            coords_eval.domain(),
            DDC_LAMBDA(Index<IDim<X, I>...> const e) { coords_eval(e) = ddc::coordinate(e); });

[...]

    // Call spline_evaluator on the same mesh we started with
    spline_evaluator_batched(spline_eval, coords_eval.span_cview(), coef.span_cview());
    `

Sure, I could write `coords_eval(e) = ddc::coordinate(ddc::select<IDim<I,I>(e));` but in pratical cases like the one I described in my long message, we would like to have an array of full coordinates (so I was not so out of topic after all).

And, at this point (this is related to discussion on Vector Fields), we have no tool to project a Chunk<Coordinates<X...>> so we cannot do the call if we remove the templating on CoordsDims...

This is more than just a `ddc::select` in the call after all.

Reply:
Ok let's accept the feature though I am still not yet really convinced we "need" it, not clear yet to me if we will have AoS or SoA data structures.